src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java

```diff
@@ -97,30 +133,118 @@ public void startEntry(String refId) {
 	
 	@Override
 	public void endEntry() {
+        if (!inTocEntry) {
+            throw new RuntimeException("Unexpected end of toc-entry");
+        }
 		inTocEntry = false;
 	}
 
+    @Override
+    public FormatterCoreImpl getEntryOnResumed() {
+        assertInTocEntryOnResumed();
+        return tocEntryOnResumedBlocks;
+    }
+
+    @Override
+	public void startEntryOnResumed(String range) {
+		if (inTocEntry || inTocEntryOnResumed) {
+			throw new RuntimeException("toc-entry and toc-entry-on-resumed may not be nested");
+		}
+		inTocEntryOnResumed = true;
+        hasTocEntryOnResumed = true;
+        
+        String startRefId = null;
+        String endRefId = null;
+        /**
+         * parse the range attribute
+         * the pattern matches only if the range is in one of these forms:
+         * [startRefId,endRefId] (unsupported) or [startRefId,endRefId) or [startRefId,)
+         * if it matches, it returns two groups: the first one is startRefId and
+         * the second one is endRefId followed by either a ']' or a ')' character
+         */
+        Pattern p = Pattern.compile("^\\[([^,\\[\\]\\)]+),([^,\\[\\]\\)]+\\]|[^,\\[\\]\\)]*\\))$");
+        Matcher m = p.matcher(range);
+        if (m.find()) {
+            startRefId = m.group(1).trim();
+            endRefId = m.group(2);
+            if (endRefId.endsWith("]")) {
+                throw new UnsupportedOperationException(String.format("Found range %s. Ranges in the form [startRefId,endRefId] are unsupported. Please use this form: [startRefId,endRefId)", range));
+            }
+            endRefId = endRefId.substring(0, endRefId.length() - 1).trim();
+            if (endRefId.length() == 0) {
+                endRefId = null;
+            }
+        }
+        if (startRefId == null) {
+            throw new RuntimeException(String.format("Could not parse this range: %s", range));
+        }
+        TocEntryOnResumedRange entryOnResumedRange = new TocEntryOnResumedRange(startRefId, endRefId);
+		if (rangeForResumedBlock.put(tocEntryOnResumedBlocks.getCurrentBlock(), entryOnResumedRange) != null) {
+			// note that this is not strictly forbidden by OBFL, but it simplifies the implementation
+			throw new RuntimeException("No two toc-entry-on-resumed's may be contained in the same block");
+		}
+	}
+	
+	@Override
+	public void endEntryOnResumed() {
+        if (!inTocEntryOnResumed) {
+            throw new RuntimeException("Unexpected end of toc-entry-on-resumed");
+        }
+        inTocEntryOnResumed = false;
+	}
+
 	/**
 	 * Filter out the toc-entry with a ref-id that does not satisfy the predicate. This is used to
 	 * create the volume range toc. toc-block that have all their descendant toc-entry filtered out
 	 * are also omitted.
-	 */
-	/*
-	 * Note that, because this is implemented by filtering a fixed sequence of blocks, and because
+	 *
+	 * <p>Note that, because this is implemented by filtering a fixed sequence of blocks, and because
```

Comment:
If we add a paragraph start tag, we should have an end tag as well to make uniform code.

Reply:
Done.